shell/browser/serial/serial_chooser_context.cc

```diff
@@ -0,0 +1,166 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "shell/browser/serial/serial_chooser_context.h"
+
+#include <utility>
+
+#include "base/base64.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/values.h"
+#include "content/public/browser/device_service.h"
+#include "mojo/public/cpp/bindings/pending_remote.h"
+
+namespace electron {
+
+constexpr char kPortNameKey[] = "name";
+constexpr char kPersistentIdKey[] = "persistent_id";
+constexpr char kTokenKey[] = "token";
+
+std::string EncodeToken(const base::UnguessableToken& token) {
+  const uint64_t data[2] = {token.GetHighForSerialization(),
+                            token.GetLowForSerialization()};
+  std::string buffer;
+  base::Base64Encode(
+      base::StringPiece(reinterpret_cast<const char*>(&data[0]), sizeof(data)),
+      &buffer);
+  return buffer;
+}
+
+base::UnguessableToken DecodeToken(base::StringPiece input) {
+  std::string buffer;
+  if (!base::Base64Decode(input, &buffer) ||
+      buffer.length() != sizeof(uint64_t) * 2) {
+    return base::UnguessableToken();
+  }
+
+  const uint64_t* data = reinterpret_cast<const uint64_t*>(buffer.data());
+  return base::UnguessableToken::Deserialize(data[0], data[1]);
+}
+
+base::Value PortInfoToValue(const device::mojom::SerialPortInfo& port) {
+  base::Value value(base::Value::Type::DICTIONARY);
+  if (port.display_name && !port.display_name->empty())
+    value.SetStringKey(kPortNameKey, *port.display_name);
+  else
+    value.SetStringKey(kPortNameKey, port.path.LossyDisplayName());
+  if (SerialChooserContext::CanStorePersistentEntry(port))
+    value.SetStringKey(kPersistentIdKey, port.persistent_id.value());
+  else
+    value.SetStringKey(kTokenKey, EncodeToken(port.token));
+  return value;
+}
+
+SerialChooserContext::SerialChooserContext() = default;
+SerialChooserContext::~SerialChooserContext() = default;
+
+void SerialChooserContext::GrantPortPermission(
+    const url::Origin& requesting_origin,
+    const url::Origin& embedding_origin,
+    const device::mojom::SerialPortInfo& port) {
+  base::Value value = PortInfoToValue(port);
+  port_info_.insert({port.token, value.Clone()});
+
+  ephemeral_ports_[{requesting_origin, embedding_origin}].insert(port.token);
```

Comment:
when do we remove entries from `ephemeral_ports_`?

Reply:
Looking at the chromium implementation: https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/serial/serial_chooser_context.cc?q=ephemeral_ports_&ss=chromium, it doesn't look they ever get removed - maybe because they are a list of ports  to which an origin has access to/been granted permission to:
https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/serial/serial_chooser_context.h;l=92